### ISSUE TITLE
[Help]: Error when organize_files_tops_linux.csh - interpolation point outside of data constraints

### ISSUE THREAD

**Issue by anaferreira97** [gmtsar/gmtsar#445]
Hello! I'm trying to process 5 s1 images with descending pass. I tried to do a full 35 images analysis however It gave me interpolation errors and I'm using this image quantity for test purposes:

OS: Ubuntu 22.04
GMTSAR version: 6.2.2

I have 5 slc images with descending pass that i used to test the gmtsar processing. When i run:

`organize_files_tops_linux.csh SAFE_filelist ../reframed/pins.ll 1`
`organize_files_tops_linux.csh SAFE_filelist ../reframed/pins.ll 2`

I get the following errors throughout: 

```
Reading in 5862 lines of info from XML...

Writing 17 lines for orbit...

Writing 58 lines of precise orbit for the LED file...

interpolation point outside of data constraints, 3369558.888874 3393522.000000 3394092.000000

SKIP 20200209, as it does not have enough scenes ...
```

Full log file:

[organize_log.txt](https://github.com/gmtsar/gmtsar/files/9618517/organize_log.txt)

SLC EXTENT:  

![Captura de ecrã 2022-09-21 155335](https://user-images.githubusercontent.com/98409351/191557670-c033b074-c0c8-4a03-8a76-f2c785f2d9a6.png)

![slc_extents](https://user-images.githubusercontent.com/98409351/191558492-53e29941-ed58-4901-b892-96b317b450fe.png)

DEM GMTINFO
![Captura de ecrã 2022-09-21 161030](https://user-images.githubusercontent.com/98409351/191557621-c3cd1e55-4655-4ce1-a5f1-73df631cecb3.png)

Data inside /data folder

![ls_data](https://user-images.githubusercontent.com/98409351/191558772-ba99d175-808d-4c00-9e6e-cfd2b3e4c65c.png)

I already tried to increase the AOI however It is giving me the same error.

Thank you!

**Comment by Xiaohua-Eric-Xu** [comment]
> interpolation point outside of data constraints, 3369558.888874 3393522.000000 3394092.000000

This line basically mean that the orbit does not cover the selected point and it is produce by `SAT_llt2rat` from this line below in the script:
```
set azi1 = `echo $pin1 | awk '{print $1,$2,0}' | SAT_llt2rat tmp2.PRM 1 | awk '{printf("%d",$2+0.5 + '$tmpazi')}'`
```
If you would like to debug, simply pop out the parameters with `echo`, should be simple to figure out. 
Also, there was an error associated with the AOI being very small (i.e. 1 burst, thus extracted orbit being short) but I have fixed that prior to V6.2, and based on `Writing 58 lines of precise orbit for the LED file...`, this should not be the case. Just FYI.

**Comment by anaferreira97** [comment]
@Xiaohua-Eric-Xu my pins are the following:

![pins_ll](https://user-images.githubusercontent.com/98409351/191616435-348a01a3-a4f9-44d7-bfd4-71ae30b52686.png)

**Comment by Xiaohua-Eric-Xu** [comment]
Could you edit your organize_files_tops_linux.csh script and add these two lines below line 149:
```
149         set azi1 = `echo $pin1 | awk '{print $1,$2,0}' | SAT_llt2rat tmp2.PRM 1 | awk '{printf("%d",$2+0.5 + '$tmpazi')}'`
150         echo $pin1
151          cat tmp2.PRM
```

**Comment by SteffanDavies** [comment]
@Xiaohua-Eric-Xu do you see anything wrong here?

**Comment by chintals** [comment]
> Could you edit your organize_files_tops_linux.csh script and add these lines below line 149:
> 
> ```
> 149         set azi1 = `echo $pin1 | awk '{print $1,$2,0}' | SAT_llt2rat tmp2.PRM 1 | awk '{printf("%d",$2+0.5 + '$tmpazi')}'`
> 150         echo $pin1
> 151         cat tmp2.PRM
> 152         exit 1
> ```
> 
> Then run the script again and see what pops out at the end?

I would be curious to know what the first azi location is as well since the Sentinel scene overlay does not give a good idea of burst edge locations. Negative azi1 would cause the code to skip the scene.

`echo $azi1`

The first pin lon location is set to -9.1 and the second pin lon location is set -9.0. It could be possible that the first pin is in the second subswath, while the second pin is the first subswath ( which could cause azi1 to be a negative number for subswath 1) and skip assembling the scenes.

Could try a pin.ll configuration  (assuming processing one single swath only)
-8.5 39.2
-9 38.9

**Comment by Xiaohua-Eric-Xu** [comment]
Usually the PRM should not start exactly around 12:00AM of the day, but I am not sure yet what produced these values.
```
SC_clock_start = 2020003.0000000000
SC_clock_stop = 2020003.0000000000
clock_start = 3.000000000000
clock_stop = 3.000000000000
```
If I change the time to the right 2020003.2803472222. The line in the script will go through. 
Note the tmp2.PRM/LED is produced using mode 0 for `make_s1a_tops`. So are you able to reproduce the tmp2.PRM with the following command (based on your log file)?
```
make_s1a_tops /home/jpereira/ssdarray/sbas/processing/dsc/data/S1A_IW_SLC__1SDV_20200104T064314_20200104T064342_030647_03831F_771C.SAFE/annotation/*iw1*vv*xml /home/jpereira/ssdarray/sbas/processing/dsc/data/S1A_IW_SLC__1SDV_20200104T064314_20200104T064342_030647_03831F_771C.SAFE/measurement/*iw1*vv*tiff tmp2 0
```
And also, could you check the time in your xml file and see whether it got contaminated somehow. (location is product->imageAnnotation->imageInformation->productFirstLineUtcTime)

**Comment by anaferreira97** [comment]
@Xiaohua-Eric-Xu 

My xml time is the following:

![xml_time](https://user-images.githubusercontent.com/98409351/192010534-40732ad7-5d33-42be-a197-28286ad73245.png)

**Comment by Xiaohua-Eric-Xu** [comment]
Were you able to produce tmp2.PRM that has `SC_clock_start = 2020003.0000000000`, with the command I sent above?

**Comment by mobigroup** [comment]
@anaferreira97 This notebook [test.ipynb.zip](https://github.com/gmtsar/gmtsar/files/9641468/test.ipynb.zip) produces the coherency map below:

<img width="1543" alt="image" src="https://user-images.githubusercontent.com/7342379/192158714-4c8fb66d-9965-41e5-93a7-0470569cf611.png">

and the interferograms:
<img width="994" alt="image" src="https://user-images.githubusercontent.com/7342379/192158925-60351f88-b8f4-4e05-8e83-f84121663872.png">

I copied the scene names from your screenshot above so please check that we use the same datasets.

**Comment by anaferreira97** [comment]
> Were you able to produce tmp2.PRM that has `SC_clock_start = 2020003.0000000000`, with the command I sent above?

Yes, 

![tmp2_prm](https://user-images.githubusercontent.com/98409351/192235990-e65302b1-88db-48d9-8a1d-b3b5afe8a52d.png)

**Comment by anaferreira97** [comment]
@mobigroup Yes those scene dates appear correct!

**Comment by mobigroup** [comment]
@anaferreira97 nice! Maybe do you need some help to complete your processing? 
@Xiaohua-Eric-Xu As PyGMTSAR processing for the case works there is an error in your shell scripts only (because we use the same binaries).

**Comment by Xiaohua-Eric-Xu** [comment]
> @mobigroup Yes those scene dates appear correct!

OK, there is only one line of code between what you did and the line with error
```
144         make_s1a_tops $f1/annotation/*iw1*"$org_mod"*xml $f1/measurement/*iw1*"$org_mod"*tiff tmp2 0
145         ext_orb_s1a tmp2.PRM $orbit tmp2
146         set tmpazi = `echo $pin1 | awk '{print $1,$2,0}' | SAT_llt2rat tmp2.PRM 1 | awk '{printf("%d",$2+0.5)}'`
```
If you run `ext_orb_s1a tmp2.PRM S1A_OPER_AUX_POEORB_OPOD_20210316T192029_V20200103T225942_20200105T005942.EOF tmp2`, did the tmp2.PRM change its values? 
I know it shouldn't as the c-code opens it as readable only. Just curious how is it possible that all the decimal numbers are gone.

**Comment by mobigroup** [comment]
> Just curious how is it possible that all the decimal numbers are gone.

Why do you surprised when the code rounding the values and returns integer numbers?
```
printf "%d" 1.11+0.5 
1
```

PyGMTSAR version doesn't use the rounding here (and we the simpler code supposing azimuth is zero after the shift):
```
self.make_s1a_tops(subswath, date, debug)

prm = PRM.from_file(old_filename+'.PRM')
azi1 = prm.SAT_llt2rat([pins[0], pins[1], 0], precise=1)[1]
azi2 = prm.SAT_llt2rat([pins[2], pins[3], 0], precise=1)[1]
#print ('azi1', azi1, 'azi2', azi2)
prm.shift_atime(azi1, inplace=True).update()

# Working on bursts covering $azi1 ($ll1) - $azi2 ($ll2)...
#print ('assemble_tops', subswath, date, azi1, azi2, debug)
self.assemble_tops(subswath, date, azi1, azi2, debug)
```
As I remember when I checked your code I didn't find any reason to have rounding here... Maybe could you explain it?

**Comment by anaferreira97** [comment]
@Xiaohua-Eric-Xu they remained the same

![ext_orb_s1a_tmp](https://user-images.githubusercontent.com/98409351/192497260-7d4306a1-2036-4a01-b2b1-f8a928701ddc.png)
![tmp2_prm_after_ext_orb](https://user-images.githubusercontent.com/98409351/192497262-c4911990-c02e-4141-ae99-1cf486ca284a.png)

**Comment by anaferreira97** [comment]
UPDATE: I used another virtual machine and the processing worked fine so I guess It has to be a local problem.

Nevertheless, thanks for all the help!

**Comment by Xiaohua-Eric-Xu** [comment]
Hmmm, this is super strange. Not that I can think of a condition where decimal numbers get wiped out... I am glad it worked out though.

**Comment by Xiaohua-Eric-Xu** [comment]
> Why do you surprised when the code rounding the values and returns integer numbers?

@mobigroup you picked on the wrong line. This line of code computes azimuth line numbers, thus needs rounding. What we were discussing was actual time in the format of fractional day (in the PRM), and it is weird that the decimal numbers are gone. And, it's not related your python implementation.

**Comment by SteffanDavies** [comment]
> Hmmm, this is super strange. Not that I can think of a condition where decimal numbers get wiped out... I am glad it worked out though.

I did the processing on my machine using the same scripts and it worked fine. I have no idea what went wrong on their computer because I assisted in performing the processing step-by-step and always encountered the same issue. However, doing the same on my computer worked fine. It might even be hardware related (RAM or CPU??), I really have no idea.

**Comment by mobigroup** [comment]
@Xiaohua-Eric-Xu 
> This line of code computes azimuth line numbers, thus needs rounding.

Sorry, I still don’t understand because assemble_tops tool reads these values as floats:

https://github.com/mobigroup/gmtsar/blob/6b3c9f1a00099c3badef2113381f20d17442d0fb/preproc/S1A_preproc/src_assembly/assemble_tops.c#L48

float azi_1, azi_2;
…
azi_1 = atof(argv[1]);
azi_2 = atof(argv[2]);

> What we were discussing was actual time in the format of fractional day (in the PRM), and it is weird that the decimal numbers are gone.

Thanks, I see.

**Comment by Xiaohua-Eric-Xu** [comment]
> Sorry, I still don’t understand because assemble_tops tool reads these values as floats:

Yeah, I don't remember why I did that (maybe related to comparisons?). The function that calls the two numbers take them as integers (i.e., edit_tree). Made a fix here: d54190b813961bf30f8e8ca6ef8272ebbdcc02f6

**Comment by mobigroup** [comment]
Nice, thanks for the clarification.
